### DISCUSSION TITLE
idFTPServer set Path Folder

### DISCUSSION THREAD

**Discussion by kholyphoenix1** [IndySockets/Indy#598] [Q&A]
Hello, 

I searched inside the properties of the 'idFTPServer' component and couldn't find where I can set the folder path for the FTP Server.
On the Indy 10 website I can't download any documentation for some reason all the files are unavailable.
Can anybody help me?

Thanks!

**Comment by rlebeau** [comment]
> I searched inside the properties of the 'idFTPServer' component and couldn't find where I can set the folder path for the FTP Server.

There is no such property.

If the `TIdFTPServer.PathProcessing` property is `ftppCustom`, the `OnCustomPathProcess` event is given the file path requested by the client, and is expected to return back a new path that can then be used in subsequent events.

If the `TIdFTPServer.PathProcessing` property is not `ftppCustom`, paths are processed by `TIdFTPServer` itself.  First, slashes in the client's requested path are normalized based on the `PathProcessing`, and then the normalized path is combined with the `TIdFTPServerContext.CurrentDir` property to produce a final path (`CurrentDir` is kept updated automatically by `TIdFTPServer` when processing `CWD` and `CDUP` commands).

Either way, then in subsequent events, like `OnStoreFile`, `OnRetreiveFile`, `OnListDirectory`, etc, you can translate the provided path to a local disk path as needed to access the actual file/folder.

You can initialize the `TIdFTPServerContext.CurrentDir` (and `TIdFTPServerContext.HomeDir`) property as needed when a new client connects/authenticates with the server.

> On the Indy 10 website I can't download any documentation for some reason all the files are unavailable.

https://www.indyproject.org/2021/02/10/links-to-old-indy-website-pages-are-currently-broken/

**Comment by kholyphoenix1** [comment]
I find it a little confusing and failing to get results. 

```
procedure TForm1.IdFTPServer1ChangeDirectory(ASender: TIdFTPServerContext;
  var VDirectory: TIdFTPFileName);
begin
  ASender.HomeDir:='C:/TempPDF/';
  ASender.CurrentDir:='C:/TempPDF/';
  VDirectory:='C:/TempPDF/';
end;

procedure TForm1.IdFTPServer1ListDirectory(ASender: TIdFTPServerContext;
  const APath: TIdFTPFileName; ADirectoryListing: TIdFTPListOutput;
  const ACmd: String; const ASwitches: String);
begin
  ASender.HomeDir:='C:/TempPDF/';
  ASender.CurrentDir:='C:/TempPDF/';
end; 

``` 

I get error when connecting using WinSCP.
Error listing directory '/C:\TempPDF'. 

If I change it to C:/TempPDF
Returns me the following error:
Yes, the directory exists 

```
Disconnected from server
Unable to get directory listing
Error listing directory 'C:/TempPDF'. 
```

**Comment by rlebeau** [comment]
Well, for starters, you should not be changing the value of the `HomeDir` and `CurrentDir` properties in those particular events to begin with.  Set the `HomeDir`/`CurrentDir` properties only after a client has logging in, such as in the `OnAfterUserLogin` event, before it starts issuing any navigation-related commands.  Then, use the `OnChangeDirectory` event only when you need to act on navigation requests, or to change a requested path into a different value.  And use the `OnListDirectory` event only to populate the `ADirectoryListing` with items belonging to the provided `APath`.

Secondly, you are not using the correct formatting for the `HomeDir`/`CurrentDir` properties anyway.  When a client logs in, you should set the `CurrentDir` to simply `'/TempPDF/'`, and then in your various events handler, if the requested path is like `'/TempPDF/<item>'` then have your code access `'C:\TempPDF\<item>'` as needed.  The `CurrentDir` is from the perspective of the client (ie, what it sees in reply to `PWD` and `CDUP` commands).

**Comment by kholyphoenix1** [comment]
Thanks a lot!
